### ISSUE TITLE
don't use delayed write-back for temporary images

### ISSUE THREAD

**Issue by jdtournier** [MRtrix3/mrtrix3#513]
Currently, the `Image` backend will always allocate a RAM buffer for images accessed read-write, and commit this buffer to storage in the destructor. This was done to avoid nasty network saturation issues when writing data out-of-order on a memory-mapped image backed by a network-based filesystem (i.e. pretty much every cluster setup...). 

Ideally, we'd like to avoid this delayed write-back for images backed on local filesystems, since the system copes just fine (typically) with scattered writes to a memory-mapped file (at least when the filesystem isn't mounted with the `sync` option, which isn't the default anyway). The problem is that it's almost impossible to detect whether the filesystem is local in a portable fashion... (if anyone has some bright ideas on that front, I'm all ears). But one situation where it might be safe to assume that the filesystem is local is for piped temporary images - on all platforms, the default location for these images is local, and in most OSs, it's even RAM-backed, the data won't even get written to disk. 

So maybe we can forego the delayed write-back for piped images, which would speed up all pipelined operations with relatively little effort. Thoughts?

**Comment by jdtournier** [comment]
I should also add that this would halve the RAM usage for pipelined data: since the temporary file will be in `/tmp`, which on a modern distribution would be mounted onto a RAM-backed `tmpfs` filesystem, writing directly to a memory-mapped file avoids the need to store both the buffer and the file in RAM simultaneously while performing the write-back... Seems pretty worthwhile to me!

**Comment by Lestropie** [comment]
Makes sense to me, I don't see any issue with it.

With regards to non-temporary images: is there anything stopping us from using a config file entry to toggle between enforced delayed write-back and memory-mapping?

**Comment by draffelt** [comment]
I agree. It's unlikely that the tmp directory will be a networked drive.
That said, you never know with a cluster environment so it might be good to
have a config option as Rob suggested

**Comment by jdtournier** [comment]
OK, bit confused here. Seems @lestropie was suggesting a config file option to avoid delayed write-back _every time_, whereas @draffelt suggests the option would apply only to temp files. I think for temp files, it should always be safe to use direct memory-mapping. There is already an option to specify a different temp file prefix if/when `/tmp` isn't adequate.

The problem with allowing direct memory-mapping via a config file option is that the chances are this will at some point backfire. Users are bound to access images over the network at some point, and end up scratching their heads when things stall, having forgotten all about this option they might have set ages ago. The delayed write-back is actually pretty quick on a local system - it's just for large files that it would be noticeable - along with the increased RAM requirements. I'd rather stick with delayed write-back unless we can reliably detect whether the filesystem is local or networked...

**Comment by Lestropie** [comment]
> Seems @lestropie was suggesting a config file option to avoid delayed write-back every time, whereas @draffelt suggests the option would apply only to temp files. I think for temp files, it should always be safe to use direct memory-mapping.

All correct / in agreement.

> The problem with allowing direct memory-mapping via a config file option is that the chances are this will at some point backfire.

Definitely a risk. It would have to be well documented in the installation, and there'd need to be an FAQ entry entitled "Why is MRtrix hanging?" (or similar). I'm not pushing hard on this one, just floating it out there.

**Comment by jdtournier** [comment]
I think we might have a way to detect network file systems: [statfs()](http://man7.org/linux/man-pages/man2/statfs.2.html). Might be worth a crack, especially given that [recent thread on the community forum](http://community.mrtrix.org/t/hcp-connectome-tutorial/144/5)...

**Comment by jdtournier** [comment]
OK, initial implementation with detection of network filesystems in #522 - I'll close this, let's discuss on the pull request.

**Comment by Lestropie** [comment]
Closing as the discussion moved to #522.

**Comment by maxpietsch** [comment]
This is just to document that there is a problem on some file systems/ cache size/ stride combinations and we should probably think about trying to detect these circumstances.

I get dismal performance using `mrconvert` on HCP diffusion datasets if the strides are reversed (from -1,2,3,4 to 4,3,2,1). Output file is on hdd mounted as `ext4 (rw,noatime,nodiratime,discard,errors=remount-ro)`.

## the input data:
````
mrinfo /ext/data.nii.gz
************************************************
Image:               "/ext/data.nii.gz"
************************************************
  Dimensions:        145 x 174 x 145 x 288
  Voxel size:        1.25 x 1.25 x 1.25 x 1
  Data strides:      [ -1 2 3 4 ]
  Format:            NIfTI-1.1 (GZip compressed)
  Data type:         32 bit float (little endian)
  Intensity scaling: offset = 0, multiplier = 1
  Transform:                    1           0           0         -90
                               -0           1           0        -126
                               -0           0           1         -72
  comments:          FSL5.0
````

## trigger cache misses

(do not use nifti because of #803!)
````
time mrconvert /ext/data.nii.gz ~/tmp/data4.mif -stride 4:1 -force
mrconvert: [WARNING] existing output files will be overwritten
mrconvert: [100%] uncompressing image "/ext/data.nii.gz"
mrconvert: [  0%] copying from "/media/mp1...5/T1w/Diffusion/data.nii.gz" to "/home/mp14/tmp/data4.mif"... ^C^C

real    2m16.308s
user    0m13.808s
sys 0m4.964s

cat /proc/loadavg
17.10 10.25 5.59 2/1268 3265884

# system is VERY slow for a minute or two
````

## set delayed_writeback to true; in lib/file/mmap.cpp
````
git apply patch

time mrconvert /ext/data.nii.gz ~/tmp/data4.mif -stride 4:1 -force
mrconvert: [WARNING] existing output files will be overwritten
mrconvert: [100%] uncompressing image "/ext/data.nii.gz"
mrconvert: [100%] copying from "/media/mp1...5/T1w/Diffusion/data.nii.gz" to "/home/mp14/tmp/data4.mif"

real    0m49.840s
user    1m27.580s
sys 0m6.904s
 cat /proc/loadavg
5.38 8.50 5.76 2/1258 3266039

# system and user 
[patch.txt](https://github.com/MRtrix3/mrtrix3/files/555533/patch.txt)

happy
````
